File: lib-segment/src/main/java/websiteschema/mpsegment/dict/DictionaryFactory.java
package websiteschema.mpsegment.dict;

import org.apache.log4j.Logger;
import websiteschema.mpsegment.conf.MPSegmentConfiguration;
import websiteschema.mpsegment.core.MPSegment;
import websiteschema.mpsegment.dict.domain.DomainDictFactory;
import websiteschema.mpsegment.dict.domain.DomainDictionary;
import websiteschema.mpsegment.tools.StringWordConverter;

import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;

public class DictionaryFactory {

    private final static Logger l = Logger.getLogger("segment");
    private final static DictionaryFactory ins = new DictionaryFactory();

    public static DictionaryFactory getInstance() {
        return ins;
    }

    private HashDictionary coreDict;
    private DomainDictFactory domainFactory;
    private boolean loadDomainDictionary = MPSegmentConfiguration.getINSTANCE().isLoadDomainDictionary();
    private boolean loadUserDictionary = MPSegmentConfiguration.getINSTANCE().isLoadUserDictionary();

    public HashDictionary getCoreDictionary() {
        return coreDict;
    }

    public void loadDictionary() {
        try {
            ArrayList<String> list = loadWordStr("websiteschema/mpsegment/dict.txt");

            String[] array = list.toArray(new String[0]);
            Arrays.sort(array);

            StringWordConverter converter = new StringWordConverter();
            for (String wordStr : array) {
                IWord word = converter.convert(wordStr);
                if (word != null) {
                    coreDict.addWord(word);
                }
            }
            array = null;
        } catch (IOException e) {
            l.error(e);
        } catch (DictionaryException e) {
            l.error(e);
        }
    }

    private ArrayList<String> loadWordStr(String dictResource) throws IOException {
        InputStream resourceStream = null;
        File file = new File(dictResource);
        if (file.exists()) {
            resourceStream = new FileInputStream(file);
        } else {
            resourceStream = DictionaryFactory.class.getClassLoader().getResourceAsStream(dictResource);
        }
        BufferedReader reader = new BufferedReader(new InputStreamReader(resourceStream, "utf-8"));

        coreDict = new HashDictionary();
        String line = reader.readLine();
        ArrayList<String> listWordStr = new ArrayList<String>();
        while (line != null) {
            listWordStr.add(line.trim());
            line = reader.readLine();
        }
        reader.close();
        reader = null;
        return listWordStr;
    }

    public void loadDomainDictionary() {
        if (loadDomainDictionary || loadUserDictionary) {
            domainFactory = DomainDictFactory.getInstance();
            domainFactory.buildDictionary();
        }
    }

    public void loadUserDictionary() {
        if (loadUserDictionary) {
            long l1 = System.currentTimeMillis();
            String userDictFile = MPSegmentConfiguration.getINSTANCE().getUserDictionaryFile();
            DomainDictionary domainDictionary = DomainDictFactory.getInstance().getDomainDictionary();
            UserDictionaryLoader userDictionaryLoader = new UserDictionaryLoader(domainDictionary, coreDict);
            try {
                userDictionaryLoader.loadUserDictionary(userDictFile);
                userDictionaryLoader.buildDisambiguationRule(new MPSegment());
            } catch (DictionaryException e) {
                l.error(e);
            }
            userDictionaryLoader.clear();
            l1 = System.currentTimeMillis() - l1;
            l.debug((new StringBuilder()).append("[Segment] ").append("loading user dictionary time used(ms): ").append(l1).toString());
        }
    }
}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/dict/DictionaryFactory.scala
package websiteschema.mpsegment.dict

import domain.DomainDictFactory
import websiteschema.mpsegment.concept.ConceptRepository
import websiteschema.mpsegment.conf.MPSegmentConfiguration
import websiteschema.mpsegment.tools.StringWordConverter

import io.Source
import websiteschema.mpsegment.core.MPSegment
import websiteschema.mpsegment.util.FileUtil
import scala.util.parsing.json.{JSON, JSONObject}
import java.io.InputStream

object DictionaryFactory {
  val instance = new DictionaryFactory()

  def apply() = instance
}

class DictionaryFactory {

  private val config = MPSegmentConfiguration()
  private var coreDict: HashDictionary = null
  private var englishDict: TrieDictionary = null
  private var domainFactory: DomainDictFactory = null
  private val isLoadDomainDictionary: Boolean = config.isLoadDomainDictionary()
  private val isLoadUserDictionary: Boolean = config.isLoadUserDictionary()
  private val isLoadEnglishDictionary: Boolean = config.isLoadEnglishDictionary()

  def getCoreDictionary(): HashDictionary = {
    return coreDict
  }

  def getEnglishDictionary: IDictionary = englishDict

  def loadDictionary() {
    var timestamp = System.currentTimeMillis()
    try {
      coreDict = new HashDictionary()
      val inputStream = getClass.getClassLoader.getResourceAsStream("websiteschema/mpsegment/dict.txt")

      loadDictionary(inputStream, coreDict)
    } catch {
      case e: Throwable =>
        e.printStackTrace()
    } finally {
      timestamp = System.currentTimeMillis() - timestamp
      println(s"loading dictionary time used(ms): $timestamp")
    }

  }

  def loadDictionary(inputStream: InputStream, dictionary: IDictionary) {
    val converter = new StringWordConverter()
    converter.setConceptRepository(ConceptRepository())

    loadWords(inputStream) {
      wordString: String => {
        val word = converter.convert(wordString)
        dictionary addWord word
      }
    }
  }

  private def loadWords(inputStream: InputStream)(convert : (String) => Unit) {
    val source = Source.fromInputStream(inputStream, "utf-8")

    source.getLines().foreach(wordString => {
       convert(wordString.replaceAll("(^\\[)|(,$)|(\\]$)", ""))
    })

    source.close()
  }

  def loadEnglishDictionary() {
    if (isLoadEnglishDictionary) {
      var timestamp = System.currentTimeMillis()
      englishDict = new TrieDictionary()
      val inputStream = FileUtil.getResourceAsStream(config.getEnglishDictionaryFile)
      try {
        loadDictionary(inputStream, englishDict)
      }
      catch {
        case ex: Throwable => ex.printStackTrace()
      }
      finally {
        inputStream.close()
        timestamp = System.currentTimeMillis() - timestamp
        println((new StringBuilder()).append("loading english dictionary time used(ms): ").append(timestamp).toString())
      }
    }
  }

  def loadDomainDictionary() {
    if (isLoadDomainDictionary || isLoadUserDictionary) {
      domainFactory = DomainDictFactory()
      domainFactory.buildDictionary()
    }
  }

  def loadUserDictionary() {
    if (isLoadUserDictionary) {
      var l1 = System.currentTimeMillis()
      val userDictFile = config.getUserDictionaryFile()
      val domainDictionary = domainFactory.getDomainDictionary()
      val userDictionaryLoader = UserDictionaryLoader(domainDictionary, coreDict)
      try {
        userDictionaryLoader.loadUserDictionary(userDictFile)
        userDictionaryLoader.buildDisambiguationRule(new MPSegment(MPSegmentConfiguration()))
      } catch {
        case e: Throwable =>
          e.printStackTrace()
      }
      userDictionaryLoader.clear()
      l1 = System.currentTimeMillis() - l1
      println((new StringBuilder()).append("loading user dictionary time used(ms): ").append(l1).toString())
    }
  }
}
